README.md
# CrissCross
A Criss-Cross-Confused mod that adds all sorts of machines that feel kinda sorta vanilla-ish! Highly configurable, no RF, no EU, just fuels and recipes!


Changelog.txt
1.3.9.2:
-Fixed: Server crash with Infiniapples
1.3.9.1:
-Fixed: Crash when some items disabled
1.3.9:
-Added: Ability to completely enable/disable any parts
-Changed: Scribes are no longer picky about having the EXACT stack size for a recipe, just the minimum
-Changed: GUI for scribe - completely redesigned! :D
-Changed: Scribe is now instantaneous
-Changed: GUI for Electroextractor - also redesigned!
-Changed: GUI for Cropomator - you guessed it, redesigned!
-Changed: JEI Recipe GUI to reflect changes in newly redesigned GUIs
-Broke: How the word redesigned sounds in your head
-Fixed: Scribes eating your book if you messed up / changed the recipe
1.3.8:
-Added: Dependency on Forge 11.15.1.1722 or newer
-Added: Capabilities Compatibility to all TileEntities
-Added: JEI Progress Clickable Compatibility
-Added: Everful Bucket is now a "Fluid Container", meaning it'll (kinda) work on tanks: not perfectly
-Changed: Specialty Bucket textures
-Changed: Generalized Machines to using fluid containers instead of specific items
-Changed: GUIs for Stone & Cobblestone Generators
-Changed: Alternate GUI Textures if Forge Bucket Model is enabled
-Removed: Ability to customize Woodchipper fuels (see above with fluid containers)
-Fixed: Stone Generators accepting anything as a catalyst instead of specifically 
_____________________________________________________________________________________________________
1.3.7:
-Fixed: Water bucket consumption for cobblestone generators
_____________________________________________________________________________________________________
1.3.6:
-Added: XP quantity from Electrolytic Dust when smelted
-Changed: Config file for Electroextractor now requires full Ore Dictionary name like oreIron, oreGold
-Changed: Default Electroextractor fuel is now redstone dust, not blocks
-Changed: Default Electroextractor Ore Dictionary configs to reflect new parsing
-Changed: Electroextractor default (non-configurable) recipe for sand from cobblestone
-Changed: Refactoring in Woodchipper recipes
-Changed: Moved water bucket addition to the block generator tile classes
-Fixed: ItemStack stacksize expectation for Cropomator - no visible change, but fixes JEI recipe
_____________________________________________________________________________________________________
1.3.5:
-Fixed: world load / unload issues with machines
-Removed: support for 1.8.8, as Forge's ModelBakery has changed, and this change is not reflected in 1.8.8 builds
_____________________________________________________________________________________________________
1.3.4:
-Added: animations to JEI Recipe Lists
-Fixed: Proxy Crash
-Fixed: unlocalized name discrepancies
-Removed: atmos manipulator (temporarily) - was never completely implemented
_____________________________________________________________________________________________________
1.3.3:
-Added: JEI Integration for machines (except Cobble and Stone generators.. you know what those do)
-Added (WIP, not functional): Atmos Manipulator - changes weather, need to work on packet handling..
-Changed: Internal Refactors, shouldn't affect squat
-Changed: Scribe Recipes are configurable
-Changed: Configuration GUI - now better organized, but will require you to reconfigure the mod again
-Changed: Version Numbering. The old one (and I) was dumb.
_____________________________________________________________________________________________________
1.31:
-Changed: Pouches are now made differently; they no longer store color in Metadata, instead they use NBT, and can be any hex color!
	Note: Pouches can be dyed via Ore Dictionary Dyes! Unlike leather armor and sheep (might make a fix for that in V-Tweaks...)

-Added: Crafting Handler to address new changes to pouches.
	Take your "broken" pouch and put it in your crafting grid, you'll get a new one back, keeping all stored items
-Added: ability to clear pouch of color like vanilla armor; use a cauldron with water
-Added: the Scribe: new Tile Entity for crafting enchanted books. Only crafts Level I enchanted books; recipes and process times are non-configurable.

The Scribe only crafts enchanted books at Level I. This is for balance, since it does not require XP any other ways.

Scribe Recipes are as follows:
16 x Puffer Fish                          -> Aqua Affinity
8 x Spider Eye                            -> Bane of Arthropods
8 x Obsidian                              -> Blast Protection
16 x Prismarine Shard                     -> Depth Strider
8 x Redstone Blocks                       -> Efficiency
32 x Feather                              -> Feather Falling
8 x Blaze Rod                             -> Fire Aspect
16 x Blaze Powder                         -> Flame
12 x Lapis Block                          -> Fortune
8 x Diamond                               -> Infinity
4 x Piston                                -> Knockback
3 x Emerald                               -> Looting
16 x Prismarine Crystals                  -> Luck of the Sea
16 x Normal Fish                          -> Lure
16 x Quartz (item)                        -> Power
16 x Log (oreDict works)                  -> Projectile Protection
4 x Iron Block                            -> Protection
12 x Gold                                 -> Punch
Potion of Water Breathing (8:00 duration) -> Respiration
16 x Quartz Block                         -> Sharpness
8 x Slime Block                           -> Silk Touch
48 x Rotten Flesh                         -> Smite
64 x Rose (double tall)                   -> Thorns
24 x Iron Ingot                           -> Unbreaking

Optionally, see this link: http://puu.sh/lXDTk/34d6366e88.gif

src/main/java/com/oitsjustjose/criss_cross/lib/Config.java
package com.oitsjustjose.criss_cross.lib;

import java.io.File;
import java.util.List;

import com.google.common.collect.Lists;

import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraftforge.common.config.ConfigCategory;
import net.minecraftforge.common.config.Configuration;
import net.minecraftforge.common.config.Property;
import net.minecraftforge.fml.client.event.ConfigChangedEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.registry.GameRegistry;

public class Config
{
	public static Configuration config;
	public static ConfigCategory Cropomator;
	public static ConfigCategory Electroextractor;
	public static ConfigCategory Woodchipper;
	public static ConfigCategory OtherMachines;
	public static ConfigCategory FeatureControl;
	// Cropomator Configs
	public static int cropomatorProcessTime;
	public static int cropomatorOutput;
	public static String cropomatorRecipeItem;
	public static String[] cropomatorInputs;
	public static String[] cropomatorCatalysts;
	public static String[] defaultCropomatorInputs = new String[] { "minecraft:apple", "minecraft:reeds", "minecraft:nether_wart", "minecraft:carrot", "minecraft:cactus", "minecraft:potato", "minecraft:sapling:0", "minecraft:sapling:1", "minecraft:sapling:2", "minecraft:sapling:3", "minecraft:sapling:4", "minecraft:sapling:5", "minecraft:melon", "minecraft:brown_mushroom", "minecraft:red_mushroom", "minecraft:dye:3", "minecraft:pumpkin", "minecraft:wheat", "minecraft:wheat_seeds" };
	public static String[] defaultCropomatorCatalysts = new String[] { "minecraft:dye:15" };
	// Electroextractor Configs
	public static int electroextractorProcessTime;
	public static int electroextractorOutput;
	public static String electroextractorRecipeItem;
	public static String[] electroextractorFuels;
	public static String[] electroextractorOreDictInputs;
	public static String[] defaultelEctroextractorOreDictInputs = new String[] { "oreIron:12428902", "oreGold:16772608" };
	public static String[] defaultElectroextractorFuels = new String[] { "minecraft:redstone" };
	// Woodchipper Configs
	public static int woodchipperProcessTime;
	public static int woodchipperOutput;
	public static String woodchipperRecipeItem;
	// Scribe Configs
	public static String[] scribeRecipes;
	public static String[] defaultScribeRecipes = new String[] { "minecraft:fish:3*16=6", "minecraft:spider_eye:0*8=18", "minecraft:obsidian:0*8=3", "minecraft:prismarine_shard:0*16=8", "minecraft:feather:0*32=2", "minecraft:blaze_rod:0*8=20", "minecraft:blaze_powder:0*16=50", "minecraft:piston:0*4=19", "minecraft:prismarine_crystals:0*16=61", "minecraft:fish:0*16=62", "minecraft:potion:8269*1=5", "minecraft:slime:0*8=33", "minecraft:rotten_flesh:0*48=17", "minecraft:double_plant:4*64=7", "ingotIron*24=34", "ingotGold*12=49", "gemQuartz*16=48", "gemDiamond*8=51", "gemEmerald*3=21", "logWood*16=4", "blockIron*4=0", "blockQuartz*16=16", "blockLapis*12=35", "blockRedstone*8=32" };
	// Block Generator Configs (stone and cobble)
	public static int blockGeneratorProcessTimes;
	// Feature enabler
	public static boolean enableCropomator;
	public static boolean enableElectroextractor;
	public static boolean enableWoodchipper;
	public static boolean enableCobblestoneGenerator;
	public static boolean enableStoneGenerator;
	public static boolean enableScribe;
	public static boolean enablePouches;
	public static boolean enableMantlesmashers;
	public static boolean enableInfiniapple;
	public static boolean enableBuckets;

	public static void init(File configFile)
	{
		if (config == null)
		{
			config = new Configuration(configFile);
			loadConfiguration();
		}
	}

	private static void loadConfiguration()
	{
		Property property;

		// Cropomator
		String category = "cropomator";
		List<String> propertyOrder = Lists.newArrayList();
		Cropomator = config.getCategory(category);
		config.setCategoryRequiresMcRestart(category, true);

		property = config.get(category, "Cropomator Process Time", 150, "Amount of time for the Cropomator to complete a process", 20, Short.MAX_VALUE);
		cropomatorProcessTime = property.getInt();
		propertyOrder.add(property.getName());

		property = config.get(category, "Cropomator Recipe Items", defaultCropomatorInputs, "Items to be processed by the Cropomator. Formatted as: <modid>:<item>:<meta> or <modid>:<item>.");
		cropomatorInputs = property.getStringList();
		propertyOrder.add(property.getName());

		property = config.get(category, "Cropomator Catalyst Items", defaultCropomatorCatalysts, "Items considered as catalysts by the Cropomator. Formatted as: <modid>:<item>:<meta> or <modid>:<item>.");
		cropomatorCatalysts = property.getStringList();
		propertyOrder.add(property.getName());

		property = config.get(category, "Cropomator Output Quantity", 3);
		property.comment = "Quantity output by the Cropomator";
		cropomatorOutput = property.getInt();
		propertyOrder.add(property.getName());

		property = config.get(category, "Cropomator Crafting Item", "minecraft:diamond_block");
		property.comment = "Key crafting ingredient in Cropomator's Recipe. Formatted as: <modid>:<item>:<meta> or <modid>:<item>.";
		cropomatorRecipeItem = property.getString();
		propertyOrder.add(property.getName());

		Cropomator.setPropertyOrder(propertyOrder);

		// ElectroExtractor
		category = "electroextractor";
		propertyOrder = Lists.newArrayList();
		Electroextractor = config.getCategory(category);
		config.setCategoryRequiresMcRestart(category, true);

		property = config.get(category, "Electroextractor Process Time", 250, "Amount of time for the Electroextractor to complete a process", 20, Short.MAX_VALUE);
		electroextractorProcessTime = property.getInt();
		propertyOrder.add(property.getName());

		property = config.get(category, "Electroextractor OreDictionary Entries", defaultelEctroextractorOreDictInputs, "Formatted as <OreDictName>:<DecimalColorYouWant>." + "\n" + "Use http://www.colorpicker.com/ to find a color and http://bit.ly/1RhPhcX to convert it from Hex to Decimal.");
		electroextractorOreDictInputs = property.getStringList();
		propertyOrder.add(property.getName());

		property = config.get(category, "Electroextractor's Fuel Items", defaultElectroextractorFuels, "Items considered as fuel for the Electroextractor. Formatted as: <modid>:<item>:<meta> or <modid>:<item>.");
		electroextractorFuels = property.getStringList();
		propertyOrder.add(property.getName());

		property = config.get(category, "Electroextractor Output Quantity", 2);
		property.comment = "Quantity output by the Electroextractor";
		electroextractorOutput = property.getInt();
		propertyOrder.add(property.getName());

		property = config.get(category, "Electroextractor Crafting Item", "minecraft:diamond_block");
		property.comment = "Key crafting ingredient in Electroextractor's Recipe. Formatted as: <modid>:<item>:<meta> or <modid>:<item>.";
		electroextractorRecipeItem = property.getString();
		propertyOrder.add(property.getName());

		Electroextractor.setPropertyOrder(propertyOrder);

		category = "woodchipper";
		propertyOrder = Lists.newArrayList();
		Woodchipper = config.getCategory(category);
		config.setCategoryRequiresMcRestart(category, true);

		property = config.get(category, "Woodchipper Process Time", 400, "Amount of time for the Woodchipper to complete a process", 20, Short.MAX_VALUE);
		woodchipperProcessTime = property.getInt();
		propertyOrder.add(property.getName());

		property = config.get(category, "Woodchipper Output Quantity", 6, "Amount of planks yielded from a log", 0, 64);
		woodchipperOutput = property.getInt();
		propertyOrder.add(property.getName());

		property = config.get(category, "Woodchipper Crafting Item", "minecraft:diamond_block");
		property.comment = "Key crafting ingredient in Woodchipper's Recipe. Formatted as: <modid>:<item>:<meta> or <modid>:<item>.";
		woodchipperRecipeItem = property.getString();
		propertyOrder.add(property.getName());

		Woodchipper.setPropertyOrder(propertyOrder);

		// Everything Else
		category = "other machines";
		propertyOrder = Lists.newArrayList();
		OtherMachines = config.getCategory(category);
		config.setCategoryRequiresMcRestart(category, true);

		property = config.get(category, "Scribe Recipes", defaultScribeRecipes, "Formatted as <modid>:<item>:<metadata>*<qty>=<enchantment ID>" + "\n" + "Also can be <oreDict>*qty=<enchantment ID>");
		scribeRecipes = property.getStringList();
		propertyOrder.add(property.getName());

		property = config.get(category, "Block Generators' Process Times", 20, "Amount of time for Stone and Cobble Generators to produce a block", 20, Short.MAX_VALUE);
		blockGeneratorProcessTimes = property.getInt();
		propertyOrder.add(property.getName());

		OtherMachines.setPropertyOrder(propertyOrder);

		// Feature Control
		category = "feature control";
		propertyOrder = Lists.newArrayList();
		FeatureControl = config.getCategory(category);
		config.setCategoryRequiresMcRestart(category, true);
		config.setCategoryComment(category, "Enable / Disable Features");

		property = config.get(category, "Cropomator", true);
		enableCropomator = property.getBoolean();
		propertyOrder.add(property.getName());

		property = config.get(category, "Electroextractor", true);
		enableElectroextractor = property.getBoolean();
		propertyOrder.add(property.getName());

		property = config.get(category, "Woodchipper", true);
		enableWoodchipper = property.getBoolean();
		propertyOrder.add(property.getName());

		property = config.get(category, "CobblestoneGenerator", true);
		enableCobblestoneGenerator = property.getBoolean();
		propertyOrder.add(property.getName());

		property = config.get(category, "StoneGenerator", true);
		enableStoneGenerator = property.getBoolean();
		propertyOrder.add(property.getName());

		property = config.get(category, "Scribe", true);
		enableScribe = property.getBoolean();
		propertyOrder.add(property.getName());

		property = config.get(category, "Pouches", true);
		enablePouches = property.getBoolean();
		propertyOrder.add(property.getName());

		property = config.get(category, "MantleSmashers", true);
		enableMantlesmashers = property.getBoolean();
		propertyOrder.add(property.getName());

		property = config.get(category, "Infiniapple", true);
		enableInfiniapple = property.getBoolean();
		propertyOrder.add(property.getName());

		property = config.get(category, "SpecialBuckets", true);
		enableBuckets = property.getBoolean();
		propertyOrder.add(property.getName());

		FeatureControl.setPropertyOrder(propertyOrder);

		if (config.hasChanged())
			config.save();

	}

	@SubscribeEvent
	public void onConfigurationChangedEvent(ConfigChangedEvent.OnConfigChangedEvent event)
	{
		if (event.modID.equalsIgnoreCase(Lib.MODID))
			loadConfiguration();
	}

	public static ItemStack findItemStack(String modid, String name)
	{
		if (GameRegistry.findItem(modid, name) != null)
			return new ItemStack(GameRegistry.findItem(modid, name), 1);
		else if (GameRegistry.findBlock(modid, name) != null)
			if (Item.getItemFromBlock(GameRegistry.findBlock(modid, name)) != null)
				return new ItemStack(Item.getItemFromBlock(GameRegistry.findBlock(modid, name)), 1);

		return null;
	}
}
